server/lib/actors/hot.go

```diff
@@ -0,0 +1,741 @@
+func (b *hotbucket) add(delta *hotbucket) *hotbucket {
+	if delta != nil {
+		b.total.addEntries(delta.total.entries)
+		for parent, l := range delta.groups {
+			cl, ok := b.groups[parent]
+			if ok {
+				cl.addEntries(l.entries)
+			} else {
+				b.groups[parent] = cl.dup()
+			}
+		}
+	}
+	return b
+}
+
+func (b *hotbucket) sort(btyp proto.TopicListRequest_BucketType) *hotbucket {
+	b.total.sort(btyp)
+	for _, l := range b.groups {
+		l.sort(btyp)
+	}
+	return b
+}
+
+//truncate bucket with specified size.
+func (b *hotbucket) truncate(size int) *hotbucket {
+	if size > len(b.total.entries) {
+		size = len(b.total.entries)
+	}
+	br := NewHotBucketWithLastUpdate(len(b.total.entries), b.lastUpdate)
+	for _, e := range b.total.entries[0:size] {
+		br.addEntry(&e, b.idParentMap[e.Id])
+	}
+	br.lastUpdate = b.lastUpdate
+	return br
+}
+
+func (b *hotbucket) rangeOf(ids []proto.UUID, btyp proto.TopicListRequest_BucketType, 
+	startScore int32, startId proto.UUID, count uint32) ([]HotEntry, error) {
+	if ids == nil || len(ids) <= 0 {
+		return b.total.rangeOf(btyp, startScore, startId, count), nil
+	} else if len(ids) == 1 {
+		l, ok := b.groups[ids[0]]
+		if ok {
+			return l.rangeOf(btyp, startScore, startId, count), nil
+		} else {
+			return nil, fmt.Errorf("no such id: %v", ids[0])
+		}
+	} else {
+		return nil, fmt.Errorf("ranged summeries of multiple channel not supported: %v", ids)
+	}
+}
+
+func (b *hotbucket) dup() *hotbucket {
+	br := &hotbucket {
+		groups: make(map[proto.UUID]*hotlist),
+		total: b.total.dup(),
+		idParentMap: make(map[proto.UUID]proto.UUID),
+		lastUpdate: b.lastUpdate,
+	}
+	for id, l := range b.groups {
+		br.groups[id] = l.dup()
+	}
+	for id, parent := range b.idParentMap {
+		br.idParentMap[id] = parent
+	}
+	return br
+}
+
+func (b *hotbucket) markUpdate(t time.Time) *hotbucket {
+	b.lastUpdate = t
+	return b
+}
+
+
+//
+// hotbucketStore
+//
+type hotbucketStore struct {
+	lengths []int  //how many bucket per hour/day
+	cacheLen int
+	caches []fetchCache //raw result FIFO cache of each query. smaller index has recent result
+	//summeries cache
+	risings []*hotbucket //acquire most point in during a hour/day
+	hots []*hotbucket //acquire most point among all objects which is created in duraing a hour/day/week/alltime
+	flames []*hotbucket //40~60% upvote and 
+}
+
+type fetchCache struct {
+	fetchAt time.Time
+	results []FetchResult
+}
+
+func NewHotBucketStore(lengths []int) *hotbucketStore {
+	csize := 1
+	for _, l := range lengths {
+		csize *= l
+	}
+	bs := &hotbucketStore{
+		lengths: lengths,
+		cacheLen: 0,
+		caches: make([]fetchCache, 0, csize), 
+		risings: make([]*hotbucket, proto.TopicListRequest_Day), 
+		hots: make([]*hotbucket, proto.TopicListRequest_AllTime),
+		flames: make([]*hotbucket, proto.TopicListRequest_AllTime),
+	}
+	return bs
+}
+
+func (bs *hotbucketStore) summeryBucketOf(btyp proto.TopicListRequest_BucketType, typ proto.TopicListRequest_QueryType) *hotbucket {
```

Comment:
__[gometalinter]__
* [gotype] undeclared name: proto
* [gocyclo] cyclomatic complexity 12 of function (*hotbucketStore).summeryBucketOf() is high (> 10)